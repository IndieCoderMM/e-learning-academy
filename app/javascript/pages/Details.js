import React from 'react';
import { useParams } from 'react-router-dom';
import { useSelector, useDispatch } from 'react-redux';
import axios from 'axios';

const Details = () => {
  const { id } = useParams();
  const courses = useSelector((state) => state.courses.courses);
  const course = courses.find((c) => c.id === parseInt(id, 10));
  const dispatch = useDispatch();

  const handleDelete = () => {
    axios
      .delete(`/api/courses/${id}`)
      .then(() => {
        dispatch({ type: 'DELETE_COURSE', payload: id });
      });
  };

  if (!course) {
    return <div>Course not found.</div>;
  }

  return (
    <div className="course-detail">
      <h2>{course.title}</h2>
      <img src={course.img_url} alt={course.title} className="course-detail__image" />
      <p>{course.description}</p>
      <p>
        Price: $
        {course.price}
      </p>
      <p>
        Duration:
        {' '}
        {course.duration}
        {' '}
        minutes
      </p>
      <p>
        Instructor:
        {' '}
        {course.instructor}
      </p>
      <button type="button" onClick={() => handleDelete}>Delete</button>
    </div>
  );
};

export default Details;
