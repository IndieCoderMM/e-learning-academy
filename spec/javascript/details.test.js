import React from 'react';
import { render, screen } from '@testing-library/react';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import coursesReducer from '../../app/javascript/store/coursesSlice';
import Details from '../../app/javascript/pages/Details';

const course = {
  id: 1,
  title: 'Course 1',
  description: 'Description for course 1',
  instructor: 'Instructor 1',
  price: 5.99,
  duration: 300,
  img_url: 'image1.jpg',
};

const store = configureStore({
  reducer: {
    courses: coursesReducer,
  },
  preloadedState: {
    courses: {
      courses: [course],
      loading: false,
      error: null,
    },
  },
});

describe('Details Page', () => {
  it('displays course details when course exists', () => {
    const { asFragment } = render(
      <MemoryRouter initialEntries={[`/details/${course.id}`]}>
        <Provider store={store}>
          <Routes>
            <Route path="/details/:id" element={<Details />} />
          </Routes>
        </Provider>
      </MemoryRouter>,
    );

    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByText(course.title)).toBeInTheDocument();
    expect(screen.getByText(course.description)).toBeInTheDocument();
    expect(screen.getByText(`Price: $${course.price}`)).toBeInTheDocument();
  });

  it('displays "Course not found." when course does not exist', () => {
    const courseId = '3';
    render(
      <MemoryRouter initialEntries={[`/details/${courseId}`]}>
        <Provider store={store}>
          <Routes>
            <Route path="/details/:id" element={<Details />} />
          </Routes>
        </Provider>
      </MemoryRouter>,
    );

    expect(screen.getByText('Course not found.')).toBeInTheDocument();
  });
});
