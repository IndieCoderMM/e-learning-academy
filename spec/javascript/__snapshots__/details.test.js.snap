// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details Page displays course details when course exists 1`] = `
<DocumentFragment>
  <div
    class="course-detail"
  >
    <div
      class="course-detail__image-container"
    >
      <img
        alt="Course 1"
        class="course-detail__image"
        src="image1.jpg"
      />
    </div>
    <div
      class="course-detail__details"
    >
      <h2>
        Course 1
      </h2>
      <p>
        Description for course 1
      </p>
      <p>
        Price: $5.99
      </p>
      <p>
        Duration:300 minutes
      </p>
      <p>
        Instructor:Instructor 1
      </p>
    </div>
  </div>
</DocumentFragment>
`;
